gcc/rust/backend/rust-mangle.cc

```diff
@@ -279,26 +314,155 @@ v0_add_identifier (std::string &mangled, const std::string &identifier)
   std::replace (punycode.begin (), punycode.end (), '-', '_');
 
   if (!is_ascii_ident)
-    mangled.append ("u");
+    mangled << "u";
 
-  mangled += std::to_string (punycode.size ());
+  mangled << std::to_string (punycode.size ());
   // If the first character of the identifier is a digit or an underscore, we
   // add an extra underscore
   if (punycode[0] == '_')
-    mangled += "_";
+    mangled << "_";
 
-  mangled += punycode;
+  mangled << punycode;
+  return mangled.str ();
 }
 
 static std::string
-v0_type_prefix (const TyTy::BaseType *ty)
+v0_path (const TyTy::BaseType *ty, const Resolver::CanonicalPath &path)
 {
-  auto ty_prefix = v0_simple_type_prefix (ty);
-  if (!ty_prefix.empty ())
-    return ty_prefix;
+  auto mappings = Analysis::Mappings::get ();
 
-  // FIXME: We need to fetch more type prefixes
-  rust_unreachable ();
+  std::string mangled;
+
+  path.iterate_segs ([&] (const Resolver::CanonicalPath &seg) {
+    std::string prefix = "";
+    std::string ns = "";
+    std::string ident = "";
+    std::string generic_prefix = "";
+    std::string generic_postfix = "";
+    std::string crate_disambiguator = "";
+
+    HirId hir_id;
+    rust_assert (mappings->lookup_node_to_hir (seg.get_node_id (), &hir_id));
+
+    HirId parent_impl_id = UNKNOWN_HIRID;
+    HIR::ImplItem *impl_item
+      = mappings->lookup_hir_implitem (hir_id, &parent_impl_id);
+    HIR::TraitItem *trait_item = mappings->lookup_hir_trait_item (hir_id);
+    HIR::Item *item = mappings->lookup_hir_item (hir_id);
+
+    if (impl_item != nullptr)
+      {
+	switch (impl_item->get_impl_item_type ())
+	  {
+	    case HIR::ImplItem::FUNCTION: {
+	      prefix = "N";
+	      ns = "v";
+	      HIR::Function *fn = static_cast<HIR::Function *> (impl_item);
+	      if (!fn->get_generic_params ().empty ())
+		{
+		  generic_prefix = "I";
+		  generic_postfix = v0_generic_arg (ty) + "E";
+		}
+	    }
```

Comment:
for example here you could factor this into a function, which would take an `HIR::Function&` as argument and do the required handling